File: patterns-use-cases/ticket-reservation/ticket-reservation-python/example/ticket_object.py
from restate.context import ObjectContext
from restate.object import VirtualObject


ticket = VirtualObject("ticket")


@ticket.handler()
async def reserve(ctx: ObjectContext) -> bool:
    status = await ctx.get("status") or "AVAILABLE"

    if status == "AVAILABLE":
        ctx.set("status", "RESERVED")
        return True
    else:
        return False


@ticket.handler()
async def unreserve(ctx: ObjectContext):
    status = await ctx.get("status") or "AVAILABLE"

    if status != "SOLD":
        ctx.clear("status")


@ticket.handler()
async def mark_as_sold(ctx: ObjectContext):
    status = await ctx.get("status") or "AVAILABLE"

    if status == "RESERVED":
        ctx.set("status", "SOLD")

File: typescript/tutorials/tour-of-restate-typescript/src/part4/ticket_object.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

import * as restate from "@restatedev/restate-sdk";
import { TicketStatus } from "../auxiliary/ticket_status";

export const ticketObject = restate.object({
  name: "TicketObject",
  handlers: {
    async reserve(ctx: restate.ObjectContext) {
      const status = (await ctx.get<TicketStatus>("status")) ?? TicketStatus.Available;

      if (status === TicketStatus.Available) {
        ctx.set("status", TicketStatus.Reserved);
        return true;
      } else {
        return false;
      }
    },

    async unreserve(ctx: restate.ObjectContext) {
      const status = (await ctx.get<TicketStatus>("status")) ?? TicketStatus.Available;

      if (status !== TicketStatus.Sold) {
        ctx.clear("status");
      }
    },

    async markAsSold(ctx: restate.ObjectContext) {
      const status = (await ctx.get<TicketStatus>("status")) ?? TicketStatus.Available;

      if (status === TicketStatus.Reserved) {
        ctx.set("status", TicketStatus.Sold);
      }
    },
  },
});

export const TicketObject: typeof ticketObject = { name: "TicketObject" };
